source/simulator/assembly.cc

```diff
@@ -1146,14 +1151,18 @@ namespace aspect
               // We only need to look up values of shape functions if they
               // belong to 'our' component. They are zero otherwise anyway.
               // Note that we later only look at the values that we do set here.
-              for (unsigned int k=0; k<advection_dofs_per_cell; ++k)
+              for (unsigned int k=0, i=0; i<advection_dofs_per_cell; ++k)
```

Comment:
I find this hard to read. First, the fact that there are two related, but differently named variables is not helpful. Maybe call them `k` and `k_advection` or similar?

Second, it's hard to read because you increment `k` here, but `i` elsewhere. What if you wrote
`i += (fe_system_to_component_index(k).first == solution_component ? 1 : 0)` here (presumably before you increment `k`) to keep it all in one place?